pennylane_lightning/core/src/simulators/lightning_tensor/tensor/base/TensorBase.hpp

```diff
@@ -0,0 +1,87 @@
+// Copyright 2024 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file TensorBase.hpp
+ * Tensor base class for all backends.
+ */
+
+#pragma once
+
+#include <vector>
+
+#include "Error.hpp"
+
+namespace Pennylane::LightningTensor {
+
+/**
+ * @brief CRTP-enabled base class for tensor.
+ *
+ * @tparam Precision Floating point precision.
+ * @tparam Derived Derived class to instantiate using CRTP.
+ */
+template <class PrecisionT, class Derived> class TensorBase {
+  private:
+    std::size_t rank_;                 // A rank N tensor has N modes
+    std::size_t length_;               // Number of elements
+    std::vector<std::size_t> modes_;   // modes for contraction identify
+    std::vector<std::size_t> extents_; // Number of elements in each mode
```

Comment:
Are these variables `const` as there aren't any setter methods?

Reply:
all private variables except for `length_` are set as `const`